wallet, refactor: make DescriptorScriptPubKeyMan agnostic of internal flag

S3RK opened the pull request:
Rationale: improve consistency between `CWallet` and `DescriptorScriptPubKeyMan`; simplify `ScriptPubKeyMan` interface.

Descriptor in itself is neither internal or external. It's responsibility of a wallet to assign and manage descriptors for a specific purpose. Duplicating information about internalness of a descriptor could lead to inconsistencies and unexpected behaviour (for example misreporting keypool size).

achow101 commented:
ACK 07d2fffea9ea3589060632fdfdbf44f9d31caab6

instagibbs commented:
concept ACK

instagibbs commented on src/wallet/scriptpubkeyman.cpp:
```diff
@@ -1585,12 +1585,10 @@ std::set<CKeyID> LegacyScriptPubKeyMan::GetKeys() const
 {
     // Returns true if this descriptor supports getting new addresses. Conditions where we may be unable to fetch them (e.g. locked) are caught later
-    if (!CanGetAddresses(m_internal)) {
+    if (!CanGetAddresses(true)) {
```
Please annotate the boolean arg

S3RK replied:
Thanks. Annotated and added comment that the arg is not used

achow101 reviewed (approved):
re-ACK c9a4e0d8e2846e7eb03e8d5fd3ee18dba6f3c4a1

Only change since last review is a comment.

meshcollider reviewed (commented):
utACK c9a4e0d8e2846e7eb03e8d5fd3ee18dba6f3c4a1

S3RK committed:
refactor: remove m_internal from DescriptorSPKman

Descriptor in itself is neither internal or external.
It's responsibility of a wallet to assign and manage descriptors
for a specific purpose. Duplicating such information could lead to
inconsistencies and unexpected behaviour.

achow101 reviewed (approved):
ACK 7e329e281042051fd85acd2f9dd8f893cbefcb8d

fanquake commented:
@meshcollider @instagibbs

instagibbs commented on src/wallet/scriptpubkeyman.cpp:
```diff
@@ -1613,12 +1613,10 @@ std::set<CKeyID> LegacyScriptPubKeyMan::GetKeys() const
 {
     // Returns true if this descriptor supports getting new addresses. Conditions where we may be unable to fetch them (e.g. locked) are caught later
-    if (!CanGetAddresses(m_internal)) {
+    if (!CanGetAddresses(/* internal= */ true)) { // internal arg is not used for descriptor spk managers
```
Seems less confusing to just leave the optional argument out

S3RK replied:
I don't know how I missed that. Thanks

instagibbs reviewed (approved):
utACK

Idea for follow-up:

Extract the desc_str creation from `DescriptorScriptPubKeyMan::SetupDescriptorGeneration`, then I think the logic/flow between `ExternalSignerScriptPubKeyMan` and `DescriptorScriptPubKeyMan` becomes a lot closer, could reduce amount of required duplicate code.

It would also mean that DescriptorSPKM never have to know anything about internal-ness, just like External ones.

jb55 commented:
ACK c94e3da57e85812ad391e4f7173321781419b9c0

instagibbs commented:
reACK https://github.com/bitcoin/bitcoin/pull/20191/commits/181181019c5baa3e2d5b675d1843a45aa028781c

achow101 commented:
reACK 181181019c5baa3e2d5b675d1843a45aa028781c

fanquake commented:
Looks like there are now some CI failures  for the `wallet_importdescriptors.py --descriptors` test. An incompatiblity between this and #19651?

achow101 commented:
@fanquake Fix in #22379
